daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1304,25 +1323,37 @@ private[lf] object SBuiltin {
                 SResultNeedKey(
                   GlobalKeyWithMaintainers(gkey, keyWithMaintainers.maintainers),
                   onLedger.committers,
-                  {
-                    case SKeyLookupResult.Found(cid) =>
-                      onLedger.ptx = onLedger.ptx.copy(
-                        keys = onLedger.ptx.keys + (gkey -> KeyActive(cid)),
-                        globalKeyInputs = onLedger.ptx.globalKeyInputs + (gkey -> KeyActive(cid)),
-                      )
-                      // We have to check that the discriminator of cid does not conflict with a local ones
-                      // however we cannot raise an exception in case of failure here.
-                      // We delegate to SEImportValue the task to check cid.
-                      machine.ctrl = operation.cidToSExpr(cid)
-                      true
-                    case SKeyLookupResult.NotFound =>
-                      onLedger.ptx = onLedger.ptx.copy(
-                        keys = onLedger.ptx.keys + (gkey -> KeyInactive),
-                        globalKeyInputs = onLedger.ptx.globalKeyInputs + (gkey -> KeyInactive),
-                      )
-                      operation.handleKeyNotFound(machine)
-                    case SKeyLookupResult.NotVisible =>
-                      machine.tryHandleSubmitMustFail()
+                  result => {
+                    cacheGlobalLookup(onLedger, gkey, result)
+                    // We need to check for activeness here since we only
+                    // modify keys if the archive was for a key
+                    // already brought into scope.
```

Comment:
The comment is a bit confusing since the following lines do not use `ActiveKey` nor `InactiveKey`.

Reply:
6d05e4c41660720c4ce494d3f057612de9978261